cmd/mailbox-controller/controller.go

```diff
@@ -0,0 +1,263 @@
+		go ctl.syncLoop(ctx, worker)
+	}
+	<-doneCh
+}
+
+func (ctl *mbCtl) OnAdd(obj any) {
+	logger := klog.FromContext(ctl.context)
+	logger.V(4).Info("Observed add", "obj", obj)
+	ctl.enqueue(obj)
+}
+
+func (ctl *mbCtl) OnUpdate(oldObj, newObj any) {
+	logger := klog.FromContext(ctl.context)
+	logger.V(4).Info("Observed update", "oldObj", oldObj, "newObj", newObj)
+	if newObj != nil {
+		ctl.enqueue(newObj)
+	} else if oldObj != nil {
+		ctl.enqueue(oldObj)
+	}
+}
+func (ctl *mbCtl) OnDelete(obj any) {
+	logger := klog.FromContext(ctl.context)
+	logger.V(4).Info("Observed delete", "obj", obj)
+	ctl.enqueue(obj)
+}
+
+func (ctl *mbCtl) enqueue(obj any) {
+	logger := klog.FromContext(ctl.context)
+	switch typed := obj.(type) {
+	case *tenancyv1alpha1.Workspace:
+		logger.V(4).Info("Enqueuing reference due to workspace", "wsName", typed.Name)
+		ctl.queue.Add(typed.Name)
+	case *workloadv1alpha1.SyncTarget:
+		mbwsName := mbwsNameOfSynctarget(typed)
+		logger.V(4).Info("Enqueuing reference due to SyncTarget", "wsName", mbwsName, "syncTargetName", typed.Name)
+		ctl.queue.Add(mbwsName)
+	default:
+		logger.Error(nil, "Notified of object of unexpected type", "object", obj, "type", fmt.Sprintf("%T", obj))
+	}
+}
+
+func (ctl *mbCtl) syncLoop(ctx context.Context, worker int) {
+	doneCh := ctx.Done()
+	logger := klog.FromContext(ctx)
+	logger = logger.WithValues("worker", worker)
+	ctx = klog.NewContext(ctx, logger)
+	logger.V(4).Info("SyncLoop start")
+	for {
+		select {
+		case <-doneCh:
+			logger.V(2).Info("SyncLoop done")
+			return
+		default:
+			ref, shutdown := ctl.queue.Get()
+			if shutdown {
+				logger.V(2).Info("Queue shutdown")
+				return
+			}
+			ctl.sync1(ctx, ref)
+		}
+	}
+}
+
+func (ctl *mbCtl) sync1(ctx context.Context, ref any) {
+	defer ctl.queue.Done(ref)
+	logger := klog.FromContext(ctx)
+	logger.V(4).Info("Dequeued reference", "ref", ref)
+	retry := ctl.sync(ctx, ref)
+	if retry {
+		ctl.queue.AddRateLimited(ref)
+	} else {
+		ctl.queue.Forget(ref)
+	}
+}
+
+func (ctl *mbCtl) sync(ctx context.Context, refany any) bool {
+	logger := klog.FromContext(ctx)
+	mbwsName, ok := refany.(string)
+	if !ok {
+		logger.Error(nil, "Sync expected a string", "ref", refany, "type", fmt.Sprintf("%T", refany))
+		return false
+	}
+	parts := strings.Split(mbwsName, wsNameSep)
+	if len(parts) != 2 {
+		logger.V(3).Info("Ignoring non-mailbox workspace name", "wsName", mbwsName)
+		return false
+	}
+	byIndex, err := ctl.syncTargetIndexer.ByIndex(mbwsNameIndexKey, mbwsName)
+	if err != nil {
+		logger.Error(err, "Failed to lookup SyncTargets by mailbox workspace name", "mbwsName", mbwsName)
+		return false
+	}
+	var syncTarget *workloadv1alpha1.SyncTarget
+	if len(byIndex) == 0 {
+	} else {
+		syncTarget = byIndex[0].(*workloadv1alpha1.SyncTarget)
+		if len(byIndex) > 1 {
+			logger.Error(nil, "Impossible: more than one SyncTarget fetched from index; using the first", "mbwsName", mbwsName, "fetched", byIndex)
```

Comment:
If this shouldn't happen, why not returning with error? Do we know that in this case the first one is at all valid ?

Reply:
The caller can not handle a general error, it only needs to know whether to retry.
We do not know which is the correct one to use when there are multiple, because there should never be multiple.